plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -132,6 +130,323 @@ nonZeroArg :: (Integer -> Integer -> Integer) -> Integer -> Integer -> Evaluatio
 nonZeroArg _ _ 0 = EvaluationFailure
 nonZeroArg f x y = EvaluationSuccess $ f x y
 
+{- Note [How to add a built-in function: simple cases]
+This Notes explains how to add a built-in function and how to read definitions of existing built-in
+functions. It does not attempt to explain why things the way they are, that is explained in comments
+in relevant files (will have a proper overview doc on that, but for now you can check out this
+comment: https://github.com/input-output-hk/plutus/issues/4306#issuecomment-1003308938).
+
+In order to add a new built-in function one needs to add a constructor to 'DefaultFun' and handle
+it within the @ToBuiltinMeaning uni DefaultFun@ instance like that:
+
+    toBuiltinMeaning <Name> =
+        makeBuiltinMeaning
+            <denotation>
+            <costingFunction>
+
+'makeBuiltinMeaning' creates a Plutus builtin out of its denotation (i.e. Haskell implementation)
+and a costing function for it. Once a builtin is added, its Plutus type is kind-checked and printed
+to a golden file automatically (consult @git status@).
+
+Below we will enumerate what kind of denotations are accepted by 'makeBuiltinMeaning' without
+touching any costing stuff.
+
+1. The simplest example of an accepted denotation is a monomorphic function that takes values of
+built-in types and returns a value of a built-in type as well. For example
+
+    encodeUtf8 :: Text -> BS.ByteString
+
+You can feed 'encodeUtf8' directly to 'makeBuiltinMeaning' without specifying any types:
+
+    toBuiltinMeaning EncodeUtf8 =
+        makeBuiltinMeaning
+            encodeUtf8
+            <costingFunction>
+
+This will add the builtin, the only two things that remain are implementing costing for this
+builtin (out of the scope of this Note) and handling it within the @Flat DefaultFun@ instance
+(see Note [Stable encoding of PLC]).
+
+2. If the type of the denotation has any constrained type variables in it, all of them need to be
```

Comment:
Perhaps emphasise that this is the _Haskell_ type of the denotation and that's where the type variables are coming from (I know the denotation's a Haskell object, but it's easy to get confused when there are two different type systems in play).

Reply:
I spent several minutes trying to come up with a way to make that clarification, but it just all looks worse. So I could say "If the Haskell type of the denotation has any constrained type variables in it", but that sounds like a denotation can have a non-Haskell type, which it can't. A builtin can have a Plutus type and a Haskell type, but a denotation is always a Haskell object. Plus, it's really unambiguous here which language is meant, because Plutus Core does not have any constraints. I agree it's easy to get confused in general, but I think in this particular case things are pretty unambiguous, given that the "denotation" term is introduced have a screen before and it's explicitly stated that a denotation is a Haskell implementation.

Reply:
But if you think there's a better way of formulating that part, do write it down, I'll include it in a subsequent PR, or feel free to adjust the docs yourself.